### PR TITLE
Gracefully shutdown listener pool

### PR DESCRIPTION
**Pull request by dddddai**
Use the shutdown pattern in https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ExecutorService.html

I wonder how to determine the timeout

### CODE REVIEW

**Comment by dddddai** [comment]
Perhaps it could create a new thread to shutdown so that `stop()` would not be blocked by
`awaitTermination(long timeout, TimeUnit unit)`

File: util/src/main/java/io/kubernetes/client/informer/cache/SharedProcessor.java [hunk 1]
```diff
@@ -146,7 +147,17 @@ public void stop() {
...
+    executorService.shutdown();
+    try {
+      // Wait a while for existing tasks to terminate
+      if (!executorService.awaitTermination(1, TimeUnit.MINUTES)) {
+        // Cancel currently executing tasks
+        executorService.shutdownNow();
+      }
+    } catch (InterruptedException e) {
+      executorService.shutdownNow();
+      Thread.currentThread().interrupt();
```
Comment (yue9944882) [inline util/src/main/java/io/kubernetes/client/informer/cache/SharedProcessor.java:160]: can you elaborate the point of this line? interrupting the current thread doesnt make sense to me

**Comment by yue9944882** [comment]
> Perhaps it could create a new thread to shutdown so that stop() would not be blocked by
awaitTermination(long timeout, TimeUnit unit)

1 min-ish timeout sounds reasonable to me.. in terms of observability, it's good to have more logging for longer timeout i suppose..

File: util/src/main/java/io/kubernetes/client/informer/cache/SharedProcessor.java (hunk 1 above)
Comment (dddddai) [reply util/src/main/java/io/kubernetes/client/informer/cache/SharedProcessor.java]: Yes, there is no need to interrupt the thread here. (There is no higher level interrupt handlers so it doesn't need to reraise the InterruptedException)

I have removed it.

**Comment by dddddai** [comment]
> thanks for this, overall looks good. can you add test cases verifies that:
> 
> * the stopping workers (within the timeout) should properly block the processor's `stop` call
> * the slow controller-workers are interrupted upon timeout

I add a new test func, please take a look : )

**Comment by dddddai** [comment]
> /lgtm
> 
> 60s can be too long for running a single test, we can consider a passive clock for robustness in the future.. and plz squash

Thanks for your review, I have squashed the commits.

**Comment by dddddai** [comment]
> /lgtm
> 
> 60s can be too long for running a single test, we can consider a passive clock for robustness in the future.. and plz squash

Thanks for your review, I add a func to give the timeout, so that the test would not be too long.

File: util/src/main/java/io/kubernetes/client/informer/cache/SharedProcessor.java [hunk 2]
```diff
@@ -34,6 +35,9 @@
 
   private ExecutorService executorService;
 
+  private long timeout = 1;
+  private TimeUnit unit = TimeUnit.MINUTES;
```
Comment (yue9944882) [inline util/src/main/java/io/kubernetes/client/informer/cache/SharedProcessor.java:39]: private final Duration timeout;

File: util/src/main/java/io/kubernetes/client/informer/cache/SharedProcessor.java [hunk 3]
```diff
@@ -146,7 +150,22 @@ public void stop() {
     } finally {
       lock.writeLock().unlock();
     }
-    // TODO(yue9944882): gracefully shutdown listener pool
-    executorService.shutdownNow();
+    // Disable new tasks from being submitted
+    executorService.shutdown();
+    try {
+      // Wait a while for existing tasks to terminate
+      if (!executorService.awaitTermination(timeout, unit)) {
```
Comment (yue9944882) [inline util/src/main/java/io/kubernetes/client/informer/cache/SharedProcessor.java:157]: ...(timeof.OfMillis(),  TimeUnit.MilliSecond)

File: util/src/main/java/io/kubernetes/client/informer/cache/SharedProcessor.java [hunk 4]
```diff
@@ -146,7 +150,22 @@ public void stop() {
...
+        // Cancel currently executing tasks
+        executorService.shutdownNow();
+      }
+    } catch (InterruptedException e) {
+      executorService.shutdownNow();
+    }
+  }
+
+  /** set the time to wait for existing tasks to terminate. */
+  public void setTimeout(long timeout, TimeUnit unit) {
```
Comment (yue9944882) [inline util/src/main/java/io/kubernetes/client/informer/cache/SharedProcessor.java:167]: instead of setters, let's make a parameter in the constructor

**Pull request closed without merging**